TESTS/mbed_hal/mpu/main.cpp

```diff
@@ -0,0 +1,201 @@
+        SCB_InvalidateICache();
+#endif
+
+    __ISB();
+    __DSB();
+
+}
+
+static void call_mem(const volatile uint16_t *mem_function)
+{
+    // or the address with 1 to ensure the thumb bit is set
+    ((void (*)()) ((uint32_t)mem_function | 1))();
+}
+
+static void hard_fault_handler_test()
+{
+    fault_count++;
+    mbed_mpu_enable_ram_xn(false);
+}
+
+static void mpu_fault_test(const volatile uint16_t *mem_function)
+{
+    mbed_mpu_init();
+
+    // Verify that the mpu causes faults when executing ram
+    fault_count = 0;
+    mbed_mpu_enable_ram_xn(true);
+    call_mem(mem_function);
+    TEST_ASSERT_EQUAL(1, fault_count);
+
+    // Verify that the mpu can be turned off
+    fault_count = 0;
+    mbed_mpu_enable_ram_xn(false);
+    call_mem(mem_function);
+    TEST_ASSERT_EQUAL(0, fault_count);
+
+    // Verify that the mpu causes faults when executing ram
+    fault_count = 0;
+    mbed_mpu_enable_ram_xn(true);
+    call_mem(mem_function);
+    TEST_ASSERT_EQUAL(1, fault_count);
+
+    // Verify that free turns off the mpu
+    fault_count = 0;
+    mbed_mpu_free();
+    call_mem(mem_function);
+    TEST_ASSERT_EQUAL(0, fault_count);
+}
+
+void mpu_init_test()
+{
+    for (int i = 0; i < 10; i++) {
+        mbed_mpu_init();
+    }
+
+    mbed_mpu_free();
+}
+
+void mpu_free_test()
+{
+    mbed_mpu_init();
+
+    // Enable the MPU
+    mbed_mpu_enable_ram_xn(true);
+
+    // Free and ensure execution from RAM is allowed
+    mbed_mpu_free();
+
+    call_mem(&data_function);
+}
+
+void mpu_fault_test_data()
+{
+    mpu_fault_test(&data_function);
+}
+
+void mpu_fault_test_bss()
+{
+    bss_function = ASM_BX_LR;
+    clear_caches();
+    mpu_fault_test(&bss_function);
+}
+
+void mpu_fault_test_stack()
+{
+    uint16_t stack_function;
+
+    stack_function = ASM_BX_LR;
+    clear_caches();
+    mpu_fault_test(&stack_function);
+}
+
+void mpu_fault_test_heap()
+{
+    uint16_t *heap_function = (uint16_t*)malloc(2);
+
+    TEST_ASSERT_NOT_EQUAL(NULL, heap_function);
+    *heap_function = ASM_BX_LR;
+    clear_caches();
+    mpu_fault_test(heap_function);
+
+    free(heap_function);
+}
+
+utest::v1::status_t fault_override_setup(const Case *const source, const size_t index_of_case)
+{
+    // Save old fault handlers and replace it with a new one
+    real_hard_fault_handler = NVIC_GetVector(HARDFAULT_IRQn);
+    real_mem_fault_handler = NVIC_GetVector(MEMFAULT_IRQn);
+    NVIC_SetVector(HARDFAULT_IRQn, (uint32_t)&hard_fault_handler_test);
+    NVIC_SetVector(MEMFAULT_IRQn, (uint32_t)&hard_fault_handler_test);
+
+    return greentea_case_setup_handler(source, index_of_case);
+}
+
+utest::v1::status_t fault_override_teardown(const Case *const source, const size_t passed, const size_t failed,
+                                       const failure_t reason)
+{
+    // Restore real fault handlers
+    NVIC_SetVector(HARDFAULT_IRQn, real_hard_fault_handler);
+    NVIC_SetVector(HARDFAULT_IRQn, real_mem_fault_handler);
```

Comment:
Copy-paste error - should be `MEMFAULT_IRQn`

Reply:
Fixed